cpp/include/cudf/detail/copy_if_else.cuh

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/cudf.h>
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <cudf/detail/utilities/integer_utils.hpp>
+#include <utilities/bit_util.cuh>
+#include <utilities/cuda_utils.hpp>
+
+#include <rmm/device_scalar.hpp>
+#include <cub/cub.cuh>
+
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+
+namespace {    // anonymous
+
+template <size_type block_size, typename T, typename Filter, bool has_validity>
+__launch_bounds__(block_size)
+__global__
+void copy_if_else_kernel(  column_device_view const lhs,
+                           column_device_view const rhs,
+                           Filter filter,
+                           mutable_column_device_view out,
+                           size_type * __restrict__ const valid_count)
+{   
+   const size_type tid = threadIdx.x + blockIdx.x * blockDim.x;
```

Comment:
You could use block_size here too, and everywhere you have blockDim.x. Otherwise the blockDim.x has to be mov'd from a special register, I think... (Not a huge deal in the grand scheme of things. Just something us low-level optimizers do...)